modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java

```diff
@@ -240,4 +260,34 @@ void failIfRequestSchemaDiffersFromTxTs(HybridTimestamp txTs, int requestSchemaV
             throw new InternalSchemaVersionMismatchException();
         }
     }
+
+    private static class DiffKey {
+        private final int tableId;
+        private final int fromSchemaVersion;
+        private final int toSchemaVersion;
+
+        private DiffKey(int tableId, int fromSchemaVersion, int toSchemaVersion) {
+            this.tableId = tableId;
+            this.fromSchemaVersion = fromSchemaVersion;
+            this.toSchemaVersion = toSchemaVersion;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            DiffKey diffKey = (DiffKey) o;
+            return tableId == diffKey.tableId && fromSchemaVersion == diffKey.fromSchemaVersion
+                    && toSchemaVersion == diffKey.toSchemaVersion;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(tableId, fromSchemaVersion, toSchemaVersion);
```

Comment:
Maybe we shouldn’t use varags, but honestly calculate the hash code the old fashioned way?